components/config/prefs.rs

```diff
@@ -7,251 +7,398 @@ use crate::opts;
-            Value::Number(x) => {
-                if let Some(v) = x.as_f64() {
-                    PrefValue::Number(v)
-                } else {
-                    return Err(());
-                }
-            },
-            _ => return Err(()),
-        };
-        Ok(value)
-    }
-
-    pub fn as_boolean(&self) -> Option<bool> {
-        match *self {
-            PrefValue::Boolean(value) => Some(value),
-            _ => None,
-        }
-    }
-
-    pub fn as_string(&self) -> Option<&str> {
-        match *self {
-            PrefValue::String(ref value) => Some(&value),
-            _ => None,
-        }
-    }
-
-    pub fn as_i64(&self) -> Option<i64> {
-        match *self {
-            PrefValue::Number(x) => Some(x as i64),
-            _ => None,
-        }
-    }
-
-    pub fn as_u64(&self) -> Option<u64> {
-        match *self {
-            PrefValue::Number(x) => Some(x as u64),
-            _ => None,
-        }
-    }
-}
-
-#[derive(Clone, Debug, Deserialize, Serialize)]
-pub enum Pref {
-    NoDefault(Arc<PrefValue>),
-    WithDefault(Arc<PrefValue>, Option<Arc<PrefValue>>),
+/// A convenience macro for accessing a preference value using its static path.
+/// Passing an invalid path is a compile-time error.
+#[macro_export]
+macro_rules! get_pref {
+    ($($segment: ident).+) => {{
+        let values = $crate::prefs::pref_map().values();
+        let lock = values.read()
+            .map(|prefs| prefs $(.$segment)+.clone());
+        lock.unwrap()
+    }};
 }
 
-impl Pref {
-    pub fn new(value: PrefValue) -> Pref {
-        Pref::NoDefault(Arc::new(value))
-    }
-
-    fn new_default(value: PrefValue) -> Pref {
-        Pref::WithDefault(Arc::new(value), None)
-    }
-
-    fn from_json(data: Value) -> Result<Pref, ()> {
-        let value = PrefValue::from_json(data)?;
-        Ok(Pref::new_default(value))
-    }
-
-    pub fn value(&self) -> &Arc<PrefValue> {
-        match *self {
-            Pref::NoDefault(ref x) => x,
-            Pref::WithDefault(ref default, ref override_value) => match *override_value {
-                Some(ref x) => x,
-                None => default,
-            },
-        }
-    }
-
-    fn set(&mut self, value: PrefValue) {
-        // TODO - this should error if we try to override a pref of one type
-        // with a value of a different type
-        match *self {
-            Pref::NoDefault(ref mut pref_value) => *pref_value = Arc::new(value),
-            Pref::WithDefault(_, ref mut override_value) => *override_value = Some(Arc::new(value)),
-        }
-    }
+/// A convenience macro for updating a preference value using its static path.
+/// Passing an invalid path is a compile-time error.
+#[macro_export]
+macro_rules! set_pref {
+    ($($segment: ident).+, $value: expr) => {{
+        let values = $crate::prefs::pref_map().values();
+        let mut lock = values.write()
+            .map(|prefs| prefs $(.$segment)+.clone());
+        *lock = $value;
```

Comment:
Ditto.

Reply:
This macro is actually incorrect. I had forgotten to actually _use_ it, and now that I have, I found it doesn't work. I am fixing it now.

Reply:
Oops.